src/topology/sheaves/sheaf_condition/basis_le.lean

```diff
@@ -0,0 +1,273 @@
+   (h : ∀ i, f ≫ F.map (hU.rec (opens.le_supr B.f i)).op
+           = g ≫ F.map (hU.rec (opens.le_supr B.f i)).op) :
+   f = g :=
+begin
+  subst hU, apply (hF B.f).some.hom_ext,
+  intro V, dsimp, let i := V.unop.2.some,
+  let i1 := opens.le_supr B.f i, let i2 := V.unop.2.some_spec.hom,
+  have : (opens_le_cover_cocone B.f).ι.app V.unop = i2 ≫ i1 := rfl,
+  rw [this, op_comp, F.map_comp, ←category.assoc, h i, category.assoc],
+end
+
+/-- If `F` is a sheaf and `B` a basis, `F(U) ⟶ F(V)` for `V ∈ B, V ⊆ U` are jointly monomorphic. -/
+lemma mono_to_basis_le_of_sheaf {B F} (hB : is_basis_range B)
+  (hF : is_sheaf_opens_le_cover F) {A : C} {f g : A ⟶ F.obj (op U)}
+  (h : ∀ i : basis_le B U, f ≫ F.map i.2.hom.op = g ≫ F.map i.2.hom.op) :
+  f = g :=
+mono_to_cover_of_sheaf ⟨basis_le B U, _⟩ hF
+  (supr_basis_le_eq_self U hB) (λ V, by convert h V)
+
+/-- Technical lemma. -/
+lemma cone_opens_w (c : cone (bli B U ⋙ F))
+  {i : basis_le B U} {j : B.ι} {h : B.f j ≤ U}
+  (f : (bli B U).obj i ⟶ (idf B).obj j) :
+  c.π.app i ≫ F.map f = c.π.app ⟨j,h⟩ :=
+let f' : i ⟶ (⟨j,h⟩ : basis_le B U) := f  in  c.w f'
+
+/--
+Construct a cone over `{ F(W) | ∃ V ∈ B, W ⊆ V ⊆ U }`
+    given a cone over `{ F(V) | V ∈ B, V ⊆ U }` and that `F` is a sheaf.
+-/
+def cone_opens_le_cover_of_cone_basis_le {B} (hB : is_basis_range B)
+  (hF : F.is_sheaf_opens_le_cover) (c : cone (bli B U ⋙ F)) :
+  cone ((full_subcategory_inclusion _ : opens_le_cover (basis_le_fam B U) ⥤ opens X).op ⋙ F) :=
+begin
+  use c.X, refine ⟨λW, c.π.app (W.unop.2.some) ≫ F.map W.unop.2.some_spec.hom.op, _⟩,
+  intros W₁ W₂ _, apply mono_to_basis_le_of_sheaf W₂.unop.1 hB hF,
+  intro i, dsimp, simp only [category.id_comp, category.assoc, ←F.map_comp, ←op_comp],
+  iterate 2 {rw cone_opens_w},
+  exact i.2.trans (W₂.unop.2.some_spec.trans W₂.unop.2.some.2),
+end
+
+/-- If `F` is the sheaf, `F(U)` is the limit of all `F(V)` with `V ∈ B, V ⊆ U`. -/
+def lim_basis_le_of_sheaf {B} (hB : is_basis_range B)
+  (hF : F.is_sheaf_opens_le_cover) : lim_basis_le B F :=
+begin
+  intro U, unfold basis_le_presheaf_cone, rw basis_le_cone_eq U hB,
+  let f := cone_opens_le_cover_of_cone_basis_le U F hB hF,
+  have hU := hF (basis_le_fam B U), fsplit,
+    exact λ c, hU.some.lift (f c),
+    intros c i, abstract fac { dsimp,
+      let hi : ∃ j, B.f i ≤ basis_le_fam B U j := ⟨i, le_of_eq rfl⟩,
+      convert hU.some.fac (f c) (op ⟨B.f i,hi⟩) using 1,
+      exact (c.w hi.some_spec.hom.op).symm },
+    intros c ι h,
+    apply mono_to_cover_of_sheaf ⟨basis_le B U, _⟩ hF rfl,
+    intro i, specialize h i,
+    rwa ← lim_basis_le_of_sheaf.fac F hB hF U hU c i at h,
+end
+
+namespace sheaf_hom
+
+variables {B F G}
+
+/--
+The condition that the presheaf hom `lift` is the unique extension of `α`,
+a "presheaf hom on the basic opens".
+-/
+@[nolint has_inhabited_instance]
+structure uniq_extn_struct (α : idf B ⋙ F ⟶ idf B ⋙ G) :=
```

Comment:
Since this type can only have at most one inhabitant by definition, I'm not sure we should make it a `structure`. How about an `is_extension` predicate for sheaf homomorphisms and then state uniqueness separately?

Reply:
If you look at the definition of `is_limit` it also has at most one inhabitant and may very well be replaced with exists_unique. However by making it a structure we can give names to the fields which feels more organized and clearer compared to use `some` and `some_spec` (actually the names here mirrors those in `is_limit`; `is_limit` has universal quantifier inside its fields, but I don't think it makes a difference).